Global sandboxing setting

aalekseyev opened the pull request:
- Add a new config option --sandbox that lets user control whether or not dune sandboxes actions by default.
- The actions no longer give a boolean "sandbox" or "not sandbox", they can now additionally say "no_special_requirements". The framework allows specifying things like "this command can't work with symlinks, but it can work without sandbox or with copy-based sandbox, but we don't use that right now.
- Annotated some actions in dune with `no_sandboxing` so that all tests keep passing with `--sandbox=copy`
- Tested on ocaml-platform and made sure the compilation succeeds with `--sandbox=copy` too

aalekseyev committed:
add sandbox_config types

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
add sandboxing_preference to config file

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
propagate config to the actual logic

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
Annotate all the call sites that clearly rely on no-sandboxing.
(guided by test failures)

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
cli

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
discovered another breakage with ocaml platform

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
changelog entry

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev commented:
After talking to @diml, we agreed that we should also add to the dune language the ability to opt out or opt in to sandboxing per rule before advertising this feature to the public.

I intend to do that soon.

aalekseyev committed:
explore an idea: why not treat sandbox requirement as one of the dependencies

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
the parsing layer

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
require dune 1.11 and update the changelog

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
test

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
fix bug

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
fix test

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
add to the test

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
test preserve_file_kind

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
it's not clear what the solution is

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
this bit seems unnecessary

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

diml commented on src/build_system.ml:
```diff
@@ -1366,6 +1371,34 @@ end = struct
+             Some (Some (if Sys.win32 then Copy else Symlink))
+           else
+             Code_error.raise
+               "This rule requires sandboxing with symlinks, but that won't \
```
It seems like we should use `User_error.raise ~loc` here. Otherwise it's not clear which rule we are talking about

diml commented on src/module_compilation.ml:
```diff
@@ -46,7 +46,10 @@ let build_cm cctx ~dep_graphs ~precompiled_cmi ~cm_kind (m : Module.t) =
         && (Obj_dir.need_dedicated_public_dir obj_dir) then
-          SC.add_rule sctx ~sandbox:false ~dir
+          SC.add_rule sctx
+            (* CR-someday aalekseyev: why do we have [no_sandboxing] here? *)
```
We are just doing a syscall without calling out to an external command. We can reasonably expect that sandboxing is not necessary. Though, it seems fine to remove the argument altogether

aalekseyev replied:
I added a function `Action.is_useful_to_sandbox` to help with that.

diml commented on src/sandbox_config.ml:
```diff
@@ -0,0 +1,129 @@
+open! Stdune
+
+(* ['a t] represents a total map from [Sandbox_mode.t] to ['a] *)
```
What about calling this `Sandbox_mode.Dict.t`? This is what we do in other modules such as `Cm_kind`, `Mode`, ...

aalekseyev replied:
Thanks, that's obviously better!

aalekseyev committed:
fix alias stamp file dir sandboxing

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
sandbox aliases, fix some sandboxed actions

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
introduce Sandbox_mode.Dict.t

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
improve error message, and make code easier to follow

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
remove no_sandboxing that will become unnecessary soon

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
too late for 1.11

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
do not sandbox actions that have nothing to sandbox

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
Merge remote-tracking branch 'origin/master' into global-sandboxing-setting

aalekseyev committed:
swap changelog entries

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev commented:
I found an annoying bug: promotion doesn't work when you sandbox.
The reason is that when `Action_exec` tries to interpret the `Diff` stanza, it finds the sandboxed versions of all paths, and therefore it gets confused. For a start, `is_copied_from_source_tree` doesn't work, but I suspect that's not the only issue.

Relatedly, I suspect "optional" dependencies won't work when the sandbox mode is "copy". (the "Copy" probably requires the src to exist)

diml commented:
What about "unsandboxing" the paths when the files don't match?

For optional dependencies, are you thinking of `diff?`? This was intended to be used as follow:

```
(progn
 (run cmd ...)
 (diff? ...))
```

the idea being that `cmd` might optionally generate a correction. That scheme should work with sandboxing, although we probably need to remember to copy back the optional files from the sandbox.

aalekseyev commented:
Ah, I see. I misunderstood what "optional" meant here.

aalekseyev committed:
Merge remote-tracking branch 'origin/master' into global-sandboxing-setting

aalekseyev committed:
fix tests

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

diml commented on src/build_system.ml:
```diff
@@ -825,6 +826,7 @@ end = struct
                       in
                       let rule =
                         Pre_rule.make ~locks ~context:(Some context) ~env
+                          (* [no_sandboxing] here is necessary for some reason *)
```
This comment looks stale

aalekseyev committed:
remove stale comment, remove seemingly unnecessary no_sandboxing annotations

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
make promotion work with sandboxing

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
odoc seems fine now

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
make copy_and_add_line_directive work well with sandboxing

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
add doc

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev commented:
- I fixed promotion partially (just the path handling, not the optional target handling).
- I added something to the doc.
- I tested odoc build and indeed it seems the alias dependency failure is fixed.
- I fixed the line directive stuff.

I don't want to test the optional targets thing in this PR because it's complicated and the PR is already big.

aalekseyev committed:
merge

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

jeremiedimino committed:
doc

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Add missing .ml dependencies for expect tests

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Fix dependencies of inline tests

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

aalekseyev committed:
better error if targets are missing when you move them from sandbox

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
Merge branch 'global-sandboxing-setting' of github.com:aalekseyev/dune into global-sandboxing-setting

jeremiedimino committed:
Fix deps of ppx.exe

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

aalekseyev committed:
Merge branch 'global-sandboxing-setting' of github.com:aalekseyev/dune into global-sandboxing-setting

jeremiedimino committed:
doc

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

aalekseyev committed:
move archive_files handling to [Lib.Lib_and_module.link_flags]

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
fix another missing dependency

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
Merge branch 'global-sandboxing-setting' of github.com:aalekseyev/dune into global-sandboxing-setting

aalekseyev committed:
add DUNE_SANDBOX env var

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

jeremiedimino committed:
Remove unused ignore

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

aalekseyev committed:
improve exception

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
Merge branch 'global-sandboxing-setting' of github.com:aalekseyev/dune into global-sandboxing-setting

aalekseyev committed:
Merge remote-tracking branch 'origin/master' into global-sandboxing-setting

aalekseyev committed:
merge

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
credit jdimino

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

aalekseyev committed:
Merge remote-tracking branch 'origin/master' into global-sandboxing-setting
